### PR TITLE
Fix exception thrown from MsQuicStream writes when connection is aborted by peer.

### LINKED ISSUE #53090: [HTTP/3] ReservedFrameType_Throws test fails
## System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http3_MsQuic.ReservedFrameType_Throws
Intermittently fails with:
```
Assert.Throws() Failure\nExpected: typeof(System.Net.Quic.QuicConnectionAbortedException)\nActual:   typeof(System.Exception): unable to read request headers; unexpected end of stream.\n---- System.Exception : unable to read request headers; unexpected end of stream.
   at System.Net.Test.Common.Http3LoopbackStream.ReadRequestDataAsync(Boolean readBody) in /home/manicka/Repositories/runtime.2/src/libraries/Common/tests/System/Net/Http/Http3LoopbackStream.cs:line 179
   at System.Net.Test.Common.Http3LoopbackStream.HandleRequestAsync(HttpStatusCode statusCode, IList`1 headers, String content) in /home/manicka/Repositories/runtime.2/src/libraries/Common/tests/System/Net/Http/Http3LoopbackStream.cs:line 48
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http3.<>c__DisplayClass5_1.<<ReservedFrameType_Throws>b__2>d.MoveNext() in /home/manicka/Repositories/runtime.2/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs:line 136
--- End of stack trace from previous location ---
----- Inner Stack Trace -----
   at System.Net.Test.Common.Http3LoopbackStream.ReadRequestDataAsync(Boolean readBody) in /home/manicka/Repositories/runtime.2/src/libraries/Common/tests/System/Net/Http/Http3LoopbackStream.cs:line 179
   at System.Net.Test.Common.Http3LoopbackStream.HandleRequestAsync(HttpStatusCode statusCode, IList`1 headers, String content) in /home/manicka/Repositories/runtime.2/src/libraries/Common/tests/System/Net/Http/Http3LoopbackStream.cs:line 48
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http3.<>c__DisplayClass5_1.<<ReservedFrameType_Throws>b__2>d.MoveNext() in /home/manicka/Repositories/runtime.2/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs:line 136
--- End of stack trace from previous location ---
```

### LINKED ISSUE #58078: QUIC: Peer closing connection can result in OperationCanceledException on pending write
I found this while looking into test issues.

Here's what I see:
(a) Server tries to send a GOAWAY on the client's control stream; this write pends
(b) Client closes the connection (this races with (a), but it happens pretty consistently)
(c) Server receives PEER_SEND_COMPLETE from msquic, and completes the pending write from (a) with OperationCanceledException: "Write was canceled"

I would expect this to result in QuicConnectionAbortedException instead.

Stack trace for (c) above:

```
  !!!!! ShutdownAsync: got exception System.OperationCanceledException: Write was canceled
     at System.Net.Quic.Implementations.MsQuic.MsQuicStream.HandleEventSendComplete(State state, StreamEvent& evt) in C:\github\geoffkizer\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\MsQuicStream.cs:line 1139
     at System.Net.Quic.Implementations.MsQuic.MsQuicStream.HandleEvent(State state, StreamEvent& evt) in C:\github\geoffkizer\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\MsQuicStream.cs:line 854
     at System.Net.Quic.Implementations.MsQuic.MsQuicStream.NativeCallbackHandler(IntPtr stream, IntPtr context, StreamEvent& streamEvent) in C:\github\geoffkizer\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\MsQuicStream.cs:line 830
  --- End of stack trace from previous location ---
     at System.Net.Quic.Implementations.MsQuic.Internal.ResettableCompletionSource`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token) in C:\github\geoffkizer\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\Internal\ResettableCompletionSource.cs:line 73
     at System.Net.Quic.Implementations.MsQuic.MsQuicStream.WriteAsync(ReadOnlyMemory`1 buffer, Boolean endStream, CancellationToken cancellationToken) in C:\github\geoffkizer\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\MsQuicStream.cs:line 296
     at System.Net.Test.Common.Http3LoopbackStream.SendFrameHeaderAsync(Int64 frameType, Int32 payloadLength) in C:\github\geoffkizer\runtime\src\libraries\Common\tests\System\Net\Http\Http3LoopbackStream.cs:line 162
     at System.Net.Test.Common.Http3LoopbackStream.SendFrameAsync(Int64 frameType, ReadOnlyMemory`1 framePayload) in C:\github\geoffkizer\runtime\src\libraries\Common\tests\System\Net\Http\Http3LoopbackStream.cs:line 167
     at System.Net.Test.Common.Http3LoopbackStream.SendGoAwayFrameAsync(Int64 firstInvalidStreamId) in C:\github\geoffkizer\runtime\src\libraries\Common\tests\System\Net\Http\Http3LoopbackStream.cs:line 150
     at System.Net.Test.Common.Http3LoopbackConnection.ShutdownAsync() in C:\github\geoffkizer\runtime\src\libraries\Common\tests\System\Net\Http\Http3LoopbackConnection.cs:line 249
      System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_HttpClientHandlerTest_Headers_Http3_MsQuic.GetAsync_EmptyResponseHeader_Success [FAIL]
```

### PR DESCRIPTION
**Pull request by rzikm**
Fixes #53090, closes #58078.

As of recent MsQuic versions, when we receive `SEND_COMPLETE` event in `MsQuicStream` with Cancelled = 1, we already have all the information provided by previous MsQuic events needed to determine the cause of the cancellation and can throw an appropriate exception. There is no data race on the order of events as they are indicated by the same thread (I checked MsQuic source code for that).

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 1]
```diff
@@ -1276,24 +1276,33 @@ private static uint HandleEventSendComplete(State state, ref StreamEvent evt)
...
-                if (!canceled)
+                if (abortException == null)
                 {
                     state.SendResettableCompletionSource.Complete(MsQuicStatusCodes.Success);
                 }
                 else
                 {
                     state.SendResettableCompletionSource.CompleteException(
-                        ExceptionDispatchInfo.SetCurrentStackTrace(new OperationCanceledException("Write was canceled")));
+                        ExceptionDispatchInfo.SetCurrentStackTrace(abortException));
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1305]: Is there a reason to create the exception within the lock?
Could this whole change be reduced to just:
```suggestion
                        ExceptionDispatchInfo.SetCurrentStackTrace(ThrowHelper.GetConnectionAbortedException(state.ConnectionState.AbortErrorCode)));
```

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1305]: I originally had code dealing with stream write aborts as well, until I realized that they are handled in `HandleEventPeerRecvAborted` and will never get hit here. So if accessing `state.ConnectionState.AbortErrorCode` is fine outside of the lock, then I can simplify this.

**Review by ManickaP — APPROVED** [review]
Thanks!

**Comment by rzikm** [comment]
CI failures are https://github.com/dotnet/runtime/issues/67712 or unrelated

**Pull request merged**